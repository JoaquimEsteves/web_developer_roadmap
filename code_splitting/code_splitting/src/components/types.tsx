export interface pageProps {
    onRouteChange: (route:string) => void
}