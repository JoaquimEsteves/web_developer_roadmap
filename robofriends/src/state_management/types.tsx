import { robotProps } from "../components/cardList/robots";

export const CHANGE_SEARCH_FIELD = "CHANGE_SEARCH_FIELD";

export const REQUEST_ROBOTS_PENDING = 'REQUEST_ROBOTS_PENDING';
export const REQUEST_ROBOTS_SUCCESS = 'REQUEST_ROBOTS_SUCCESS';
export const REQUEST_ROBOTS_FAILED = 'REQUEST_ROBOTS_FAILED';

export type REQUEST_ROBOTS =
    'REQUEST_ROBOTS_PENDING' |
    'REQUEST_ROBOTS_SUCCESS' |
    'REQUEST_ROBOTS_FAILED';


export type setSearchFieldAction =  {
    type: "CHANGE_SEARCH_FIELD",
    payload: string
}

export type requestRobotsAction =  {
    type: REQUEST_ROBOTS,
    payload?: string | robotProps[],
}

export type stateRobots =  {
    isPending: boolean,
    robots: robotProps[] | [],
    error: string
}

export type stateSearch =  {
    searchField: string
}


export type AppState =  {
    searchRobots: stateSearch,
    requestRobots: stateRobots
}


export type AppProps =  {
    error: string,
    isPending: boolean,
    onRequestRobots: () => void,
    onSearchChange: (event: React.ChangeEvent<HTMLInputElement> ) => void,
    robots: robotProps[] | [],
    searchField: string,
}
