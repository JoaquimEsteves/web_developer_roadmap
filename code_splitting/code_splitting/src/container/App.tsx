import React, { Component } from 'react';
import './App.scss';
// import { pageProps } from '../components/types';

import asyncComponent from '../components/asyncComponent';

interface AppState {
  route: string
}

class App extends Component<{}, AppState> {
  constructor(_: {}) {
    super(_);
    this.state = {
      route: 'page1',
    }
  }

  onRouteChange = (incoming_route: string) => {
    this.setState({ route: incoming_route });
  }

  render() {
    const AsyncPage2 = asyncComponent(this.state.route);
    return <AsyncPage2 onRouteChange={this.onRouteChange}/>
  }
}

export default App;
