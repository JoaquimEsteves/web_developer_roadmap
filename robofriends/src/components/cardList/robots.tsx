export interface robotProps {
    id: Number; name: string; username:string; email: string;
}

export const robots: robotProps[] = [
    {
        id: 1,
        name: "Top Boy",
        username: "topBoy",
        email: 'top@bo.y'
    },
    {
        id: 2,
        name: "Top Girl",
        username: "topGirl",
        email: 'top@girl.y'
    },
    {
        id: 3,
        name: "Top Man",
        username: "topMa,",
        email: 'top@man.y'
    },
    {
        id: 4,
        name: "Top Woman",
        username: "topWoman",
        email: 'top@Woman.y'
    },
    {
        id: 5,
        name: "Top Rather Not say",
        username: "topRatherNotSay",
        email: 'top@misteriousl.y'
    },

];