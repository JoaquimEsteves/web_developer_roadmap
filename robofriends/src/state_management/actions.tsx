// import * as ai from './types';
import { Dispatch } from 'react';
import {
    setSearchFieldAction, CHANGE_SEARCH_FIELD, requestRobotsAction, REQUEST_ROBOTS_PENDING,
    REQUEST_ROBOTS_SUCCESS, REQUEST_ROBOTS_FAILED
} from './types';

export const setSearchField = (text: string): setSearchFieldAction => {
    return {
        type: CHANGE_SEARCH_FIELD,
        payload: text,
    }
};

export const ROBOTS_LOCATION = 'https://jsonplaceholder.typicode.com/users';
export const setRequestRobots = () => {
    // Dispatches an requestRobotsAction
    return async (dispatch: Dispatch<requestRobotsAction>) => {
        await dispatch({
            type: REQUEST_ROBOTS_PENDING,
        })
        await fetch(ROBOTS_LOCATION)
            .then(response => {
                return response.json();
            })
            .then(data => dispatch({
                type: REQUEST_ROBOTS_SUCCESS,
                payload: data
            }))
            .catch(error => dispatch({
                type: REQUEST_ROBOTS_FAILED,
                payload: error
            }));
    }
}