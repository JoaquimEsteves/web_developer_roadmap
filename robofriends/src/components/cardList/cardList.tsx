import React from 'react';
import { robotProps } from './robots';
import Card from './card/card';

const CardList = (robot_list: robotProps[]) => {
    const cardComponents = robot_list.map(card_prop => {
        return Card(card_prop);
    });
    return (
        <div id="cardList">
            {cardComponents}
        </div>
    );
}

export default CardList;