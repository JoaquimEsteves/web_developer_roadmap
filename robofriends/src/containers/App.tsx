import React, { Dispatch } from 'react';
import CardList from '../components/cardList/cardList';
import Header from '../components/Header/header';
import Scroll from '../components/scroll/scroll';
import { robotProps } from '../components/cardList/robots';
import ErrorBoundary from '../components/error_boundaries/error_boundary';
import { setSearchField, setRequestRobots } from '../state_management/actions';
import { AppState, AppProps } from '../state_management/types';
import { connect } from 'react-redux';
import './App.scss';

export const mapStateToProps = (state: AppState) => {
    return {
        searchField: state.searchRobots.searchField,
        robots: state.requestRobots.robots,
        isPending: state.requestRobots.isPending,
        error: state.requestRobots.error,
    }
}
export const mapDispatchToProps = (dispatch: Dispatch<any>) => {
    return {
        onSearchChange: (event: React.ChangeEvent<HTMLInputElement>) => {
            console.log(event);
            return dispatch(setSearchField(event.target.value))
        },
        onRequestRobots: () => dispatch(setRequestRobots())
    }
}



export class App extends React.Component<AppProps, AppState> {

    async componentDidMount() {
        this.props.onRequestRobots();
    }
    filterRobots = (robots: robotProps[] | []) => robots.filter((robot: robotProps) => {
        let robot_name = robot.name.toLowerCase();
        let search_field = this.props.searchField.toLowerCase();
        return robot_name.includes(search_field);
    });

    render() {
        const { onSearchChange, robots, isPending } = this.props;
        if (isPending) {
            return (
                <div className="tc">
                    <ErrorBoundary>
                        <h1>Yo, loading</h1>
                    </ErrorBoundary>
                </div>
            );
        }

        return (
            <div className="tc">
                <Header />
                <div>
                    <input
                        aria-label='Search the robots yo'
                        className="searchBox"
                        type='search' placeholder='yo'
                        onChange={onSearchChange}
                    />
                </div>
                <Scroll>
                    <ErrorBoundary>
                        {/* <CardList {this.filterRobots(robots)}/> */}
                        {CardList(this.filterRobots(robots))}
                    </ErrorBoundary>
                </Scroll>
            </div>
        );
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(App);