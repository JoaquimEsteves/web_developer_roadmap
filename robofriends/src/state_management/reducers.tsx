import * as types from "./types";
import { robotProps } from "../components/cardList/robots";

const initialStateSearch: types.stateSearch = {
    searchField: '',
};

const initialStateRobots: types.stateRobots = {
    isPending: false,
    robots: [],
    error: ''
};

export const searchRobots = (state = initialStateSearch, action: types.setSearchFieldAction) => {
    switch (action.type) {
        case types.CHANGE_SEARCH_FIELD:
            return { ...state, searchField: action.payload };
        default:
            return state;
    }
}

export const requestRobots = (state = initialStateRobots, action: types.requestRobotsAction) => {
    switch (action.type) {
        case types.REQUEST_ROBOTS_PENDING:
            return {...state, isPending:true};
        case types.REQUEST_ROBOTS_SUCCESS:
            return {...state, robots: action.payload as robotProps[], isPending:false};
        case types.REQUEST_ROBOTS_FAILED:
                return {...state, error: action.payload as string, isPending:false};
        default:
            return state;
    }
}