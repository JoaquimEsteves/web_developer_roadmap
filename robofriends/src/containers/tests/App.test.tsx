import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { robotProps } from '../../components/cardList/robots';
import { AppProps } from '../../state_management/types';
import { App } from '../App';
import '../../setupTests';

let wrapper: ShallowWrapper<any, any>;

beforeEach(() => {
    const exampleProps: AppProps = {
        error: '',
        onRequestRobots: jest.fn(),
        isPending: false,
        onSearchChange: jest.fn(),
        robots: [],
        searchField: 'yo',
    }
    wrapper = shallow(<App {...exampleProps} />);
})


it('App with no Store', () => {
    expect(wrapper.debug()).toMatchSnapshot();
})

it('App filter my stuff yo', () => {
    const yo_list: robotProps[] = [
        { 'email': 'yo', id: 0, name: 'yo', 'username': 'yo' },
        { 'email': 'yay', id: 1, name: 'yay', 'username': 'yay' },
    ];
    const exampleProps: AppProps = {
        error: '',
        onRequestRobots: jest.fn(),
        isPending: false,
        onSearchChange: jest.fn(),
        robots: yo_list,
        searchField: 'yo',
    }
    const instance = shallow(<App {...exampleProps} />).instance() as App;
    expect(instance.filterRobots(yo_list)).toEqual([{ 'email': 'yo', id: 0, name: 'yo', 'username': 'yo' }]);
})

it('App is pending branch', () => {
    const exampleProps: AppProps = {
        error: 'yo',
        onRequestRobots: jest.fn(),
        isPending: true,
        onSearchChange: jest.fn(),
        robots: [],
        searchField: 'yo',
    }
    const wrapper = shallow(<App {...exampleProps} />);
    expect(wrapper.find('h1').text().includes('Yo, loading')).toBe(true);
})