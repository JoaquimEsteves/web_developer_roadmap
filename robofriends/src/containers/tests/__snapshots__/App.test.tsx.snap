// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App with no Store 1`] = `
"<div className=\\"tc\\">
  <Header />
  <div>
    <input aria-label=\\"Search the robots yo\\" className=\\"searchBox\\" type=\\"search\\" placeholder=\\"yo\\" onChange={[Function: mockConstructor]} />
  </div>
  <Scroll>
    <ErrorBoundary>
      <div id=\\"cardList\\" />
    </ErrorBoundary>
  </Scroll>
</div>"
`;
